src/TmsType.hs

```diff
@@ -0,0 +1,88 @@
+-- |This module provides types for representing turing machines in a format suitable for the service: https://turingmachinesimulator.com/
+-- Below this format will be called Tms.
+
+module TmsType where
+
+import Data.List (intercalate)
+import Data.List.Utils (replace)
+import GHC.Unicode (isAlphaNum)
+
+import TMType
+
+-- |Type of Tms tape square action.
+--
+-- 'Leave' is leave any character unchanged.
+--
+-- 'ChangeFromTo f t' is change it from 'f' to 't'.
+--
+-- 'ChangeTo t' is change it from anything to 't'.
+data TmsTapeSquare = Leave | ChangeFromTo Char Char | ChangeTo Char
+    deriving (Eq)
+
+-- |Type of Tms tape head movement
+data TmsTapeHeadMovement = MoveLeft | Stay | MoveRight
+    deriving (Eq)
+
+instance Show TmsTapeHeadMovement where
+    show MoveLeft  = "<"
+    show Stay      = "-"
+    show MoveRight = ">"
+
+-- |Type of Tms command for one tape.
+-- TmsSingleTapeCommand (action, prevState, nextState, movement).
+newtype TmsSingleTapeCommand = TmsSingleTapeCommand (TmsTapeSquare, State, State, TmsTapeHeadMovement)
+    deriving (Eq)
+
+-- |Type of Tms command for entire Turing machine.
+newtype TmsCommand = TmsCommand [TmsSingleTapeCommand]
+    deriving (Eq)
+
+-- |Type of Tms format.
+-- Tms            (name,   init     accept     commands,     tapeAlphabets).
+newtype Tms = Tms (String, [State], [[State]], [TmsCommand], [[Char]])
+    deriving (Eq)
+
+instance Show Tms where
+    show
+        (Tms
```

Comment:
Can the utilization of a pretty-printing library (for example, https://hackage.haskell.org/package/prettyprinter) simplify this code?

Reply:
With it, I think I can prettify this block of code:
```
"name: " ++ name ++ "\n" ++
"init: " ++ mergeMultipleTapeStates initial ++ "\n" ++
"accept: " ++ intercalate ", " (map mergeMultipleTapeStates acStates) ++ "\n\n" ++
intercalate "\n\n" (map showTmsCommand commands)
```
But it will be one more dependence and the code will not be much prettier, so should I try?

Reply:
I think, yes, you should. One more dependence is not a problem)